### ISSUE TITLE
grid_sample backward pass performance scales poorly with input size

### ISSUE THREAD

**Issue by coolteemf** [pytorch/pytorch#71415]
### 🐛 Describe the bug

The issue in https://github.com/pytorch/pytorch/issues/64977#issue-995764567  seem to still be present in the 3D case.

Adapted code to reproduce:
```
import timeit

import torch


def grid_sample_test(input, grid, backward):
    if backward and grid.grad is not None:
        grid.grad.zero_()
    samples = torch.nn.functional.grid_sample(
        input,
        grid,
        mode="bilinear",
        padding_mode="border",
        align_corners=True,
    )
    m = samples.mean()
    if backward:
        m.backward()

    return samples


_input = None
_grid = None
_backward = None

if __name__ == "__main__":
    torch.manual_seed(15)
    torch.set_num_threads(1)

    N = 100
    C = 2
    repeats = 100
    D_out = 4
    H_out = 13
    W_out = 13
    dtype = torch.double
    devices = ["cpu"]
    backwards = [False, True]

    input_sizes = [(2, 30, 40), (2, 300, 40), (2, 3000, 40)]

    grid_cpu = 2.0 * torch.rand((N, D_out, H_out, W_out, 3), dtype=dtype) - 1.0

    for input_size in input_sizes:
        D_in, H_in, W_in = input_size
        input_cpu = torch.rand(
            (1, C, D_in, H_in, W_in),
            requires_grad=False,
            dtype=dtype,
        ).expand((N, -1, -1, -1, -1))

        for _backward in backwards:
            for device in devices:
                _grid = grid_cpu.clone().detach().to(device).requires_grad_(True)
                _input = input_cpu.to(device)

                t = timeit.timeit(
                    "grid_sample_test(_input, _grid, _backward)",
                    globals=globals(),
                    number=repeats,
                )
                print(
                    f"device={device:>4} backward={str(_backward):>5} input size={D_in:>4}x{H_in:>4}x{W_in:<4}: {t:5.2f}"
                )
```
Which yields:
```
device= cpu backward=False input size=   2x  30x40  :  0.31
device= cpu backward= True input size=   2x  30x40  :  0.94
device= cpu backward=False input size=   2x 300x40  :  0.37
device= cpu backward= True input size=   2x 300x40  :  2.63
device= cpu backward=False input size=   2x3000x40  :  0.42
device= cpu backward= True input size=   2x3000x40  : 13.97
```
From what I understood of the issue, it is related to having to keep gradient info on the input pixels which are used to compute the output. In the worst case, there should be 8 times the number of grid elements needed to compute the output. So, the computation of the forward or backward pass should depend only on the grid size.

I'm not sure if this is correct from a practical standpoint because I'm not familiar with PyTorch internal components.

Also, I did not really understand the point that was made in the previous issue:
`If this is the case, I wonder how easy it would be to have a special case for input not requiring gradient (which is the case I'm interested in)`
Why would the input require gradient for this operation ? The operation is not in place so the input is not modified thus its gradient is not needed ?

It looks like it would not be too complicated to use the fix implemented by @to-mi for the 2D case in the 3D case but it's my first time working on pytorch internal c++ code so I don't really know how much time that will take me.
You talked about a workaround @to-mi, could you share it so that I can train my network while I'm working on this issue ?

cc @VitalyFedyunin @ngimel @albanD @mruberry @jbschlosser @walterddr @kshitij12345 

### Versions

Collecting environment information...
PyTorch version: 1.11.0.dev20220118+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 26 2021, 20:14:08)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-46-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.2.67
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 460.91.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.1
[pip3] numpy-quaternion==2021.7.28.16.11.54
[pip3] torch==1.11.0.dev20220118+cu102
[pip3] torchaudio==0.11.0.dev20220118+cu102
[pip3] torchvision==0.12.0.dev20220118+cu102
[conda] Could not collect

**Comment by to-mi** [comment]
Indeed I only worked on the 2D case. Probably it would be possible to do similar fix for 3D, but I haven't look at the code for those.

> Also, I did not really understand the point that was made in the previous issue:
If this is the case, I wonder how easy it would be to have a special case for input not requiring gradient (which is the case I'm interested in)
Why would the input require gradient for this operation ? The operation is not in place so the input is not modified thus its gradient is not needed ?

This was the original issue. The gradient is not needed, but (at least in 2D) it was always computed anyway.

> You talked about a workaround @to-mi, could you share it so that I can train my network while I'm working on this issue ?

One workaround is to crop the input such that it only contains the elements needed for the computation. For example, if your grid only lies in a small area of the full input, just crop the input to have that area. When the input is small, the penalty from unnecessarily computing (or actually allocating memory for) the input gradient is smaller. This is, of course, a bit annoying to have to do and error prone.

**Comment by coolteemf** [comment]
> One workaround is to crop the input such that it only contains the elements needed for the computation. For example, if your grid only lies in a small area of the full input, just crop the input to have that area. When the input is small, the penalty from unnecessarily computing (or actually allocating memory for) the input gradient is smaller. This is, of course, a bit annoying to have to do and error prone.
I have partially implemented this workaround already (bounding box of max displacements) but still the backward pass remains too long.
Anyways, thank you for your answer, I'll try to fix the 3D case.
